shared/chat/conversation/input-area/suggestors/index.js

```diff
@@ -0,0 +1,370 @@
+
+    _setInactive = () => this.setState(s => (s.active ? {active: null, filter: '', selected: 0} : null))
+
+    _getWordAtCursor = () => {
+      if (this._inputRef.current) {
+        const input = this._inputRef.current
+        const selection = input.getSelection()
+        if (!selection || this._lastText === null) {
+          return null
+        }
+        const text = this._lastText
+        const upToCursor = text.substring(0, selection.start)
+        const words = upToCursor.split(/ |\n/)
+        const word = words[words.length - 1]
+        const position = {end: selection.start, start: selection.start - word.length}
+        return {position, word}
+      }
+      return null
+    }
+
+    _stabilizeSelection = () => {
+      const results = this._getResults()
+      if (this.state.selected > results.length - 1) {
+        this.setState({selected: 0})
+      }
+    }
+
+    _checkTrigger = text => {
+      this._timeoutID = setTimeout(() => {
+        // inside a timeout so selection will settle, there was a problem where
+        // desktop would get the previous selection on arrowleft / arrowright
+        const cursorInfo = this._getWordAtCursor()
+        if (!cursorInfo) {
+          return
+        }
+        const {word} = cursorInfo
+        if (this.state.active) {
+          const activeMarker = this.props.suggestorToMarker[this.state.active]
+          if (!word.startsWith(activeMarker)) {
+            // not active anymore
+            this._setInactive()
+          } else {
+            this.setState({filter: word.substring(activeMarker.length)}, this._stabilizeSelection)
+            return
+          }
+        }
+        for (let marker of Object.keys(this._markerToSuggestor)) {
+          if (word.startsWith(marker)) {
+            this.setState({active: this._markerToSuggestor[marker], filter: word.substring(marker.length)})
+          }
+        }
+      }, 0)
+    }
+
+    _validateProps = () => {
+      const {active} = this.state
+      if (!active) {
+        return
+      }
+      if (
+        !this.props.dataSources[active] ||
+        !this.props.renderers[active] ||
+        !this.props.suggestorToMarker[active] ||
+        !this.props.transformers[active]
+      ) {
+        throw new Error(
+          `AddSuggestors: invalid props for suggestor '${active}', did you miss a key somewhere?`
+        )
+      }
+    }
+
+    _move = (up: boolean) => {
+      this.setState(
+        s => {
+          if (!s.active) {
+            return null
+          }
+          const length = this._getResults().length
+          const selected = (((up ? s.selected - 1 : s.selected + 1) % length) + length) % length
+          return selected === s.selected ? null : {selected}
+        },
+        () => this._triggerTransform(this._getSelected(), false)
+      )
+    }
+
+    _onChangeText = text => {
+      this._lastText = text
+      // $FlowIssue TODO (DA) but I don't think this is an issue
```

Comment:
what does flow complain about here?

Reply:
likely we want flow 89 in first and change how  this is typed